### ISSUE TITLE
python-mode not listed

### ISSUE THREAD

**Issue by andreas-roehler** [milkypostman/melpa#2898]
While there is a recipe "python-mode" which should to deliver python-mode.el, it seems not provided.

**Comment by andreas-roehler** [comment]
As for shipped python.el,  M-x list-packages RET displays "python", which is correct, it corresponds to "(provide ...)"

**Comment by milkypostman** [comment]
How do you mean? I think we have had problems with the Python mode recipe
building. I have to constantly reset it.

**Comment by andreas-roehler** [comment]
Hmm, sorry. That was done by a volunteer, unfortunatly didn't get the error.
Maybe bzr resp. launchpad isn't that suitable? As it was thought to move to gitlab, might that solve it?

**Comment by andreas-roehler** [comment]
Restricted the recipe to python-mode.el, which is known to work standalone.

**Comment by milkypostman** [comment]
It's not the recipe that is the problem it's the repo. It turns out that we keep getting into a state where the working checkout cannot update. But we don't have code to detect this so out working directory becomes stale. I'm not sure what the fix it. But we would have to write a large amount of code for a small edge case. This only seems to happen for Python mode and vc-darcs

**Comment by andreas-roehler** [comment]
Thanks. So moving to gitlab should solve this. Will send a now recipe/PR when done.

**Comment by warsaw** [comment]
Launchpad now supports git too, so if it's just bzr being problematic, you could keep the code on LP and convert the repo to git.   If it's LP itself that's being problematic, then moving to GitLab would be fine.  The GNU Mailman project moved a few weeks ago, and we're generally quite happy with GitLab.

**Comment by vigneshsarma** [comment]
Any updates on this? 
I see the recipe has not been updated.

The LP mirror seems to be a little outdated.

**Comment by andreas-roehler** [comment]
(:name python-mode :repo "python-mode-devs/python-mode" :fetcher gitlab)

**Comment by purcell** [comment]
Changed in 32bfe70. Let me know if that helps.

**Comment by purcell** [comment]
(Thanks!)

**Comment by andreas-roehler** [comment]
Works for me, thanks!

**Comment by purcell** [comment]
:+1: great!

**Comment by warsaw** [comment]
Just FTR, `python-mode.el` predates `python.el` by a decade or two.

**Comment by andreas-roehler** [comment]
@dgutov AFAIU https://github.com/company-mode/company-mode/discussions/1308 the system executable wasn't found. Seems lately vendors only distribute python2 and python3 executables. So far not Emacs related at all. Calling something like `ln -s python3 python` may help.

**Comment by dgutov** [comment]
@andreas-roehler IIUC the problem was that, even though the user has customized `python-shell-interpreter` to use `python3`, and yet the completion code was still trying `python`.

**Comment by andreas-roehler** [comment]
@dgutov Okay, that's a  point.

**Comment by cosmicexplorer** [comment]
@warsaw: https://github.com/melpa/melpa/issues/2898#issuecomment-1039346120
> Just FTR, `python-mode.el` predates `python.el` by a decade or two.

The only independent corroboration of any involvement with any development of `python-mode.el` by a "Barry A. Warsaw" is a bzr revision on launchpad in which you suddenly become maintainer of the project in mid-2012:
```
* debian/control:
    + Install myself as Maintainer w/previous maintainers approval.
[...]
* debian/copyright: Updated to reflect upstream reality.
  * debian/emacsen-install: Symlink to the .el file from the FLAVOR
    directory so that python-mode.el is on load-path.  (Closes: #311382)
```

The separate `pymacs` project from debian maintainers appeared to have worked effectively enough to enable two distinct projects to achieve a unified user experience for at least the 8 years during which Matthias Klose maintained that build process throughout the XEmacs split. The build process for debian pymacs at that time very clearly requires a full working emacs with its own `python-mode`, and the claims of a "Barry A. Warsaw" who maintained it in the 90s were added (quite sloppily) in late 2011 by Andreas Roehler.

I posted in much greater detail on the gitlab repo that I've never seen before in my entire life: https://gitlab.com/python-mode-devs/python-mode/-/work_items/105#note_3222030664. Stefan Monnier clearly credits a "Dave Love" `<fx@gnu.org>` who assigns their copyright to the FSF for the years of 2003-2004 in the commit `6f480c7a389fc851d17375ecbc39782d3049731a` that introduced to emacs in April 12 of 2004.

Perhaps we should just ask Stefan? Because the authorship claim being made here seems to rest purely upon the strength of:
1. backdated git commits,
2. an `@python.org` email address.

and as someone who has experienced first-hand the immensely excruciating and unpleasant process of assuming good faith from python contributors, the second isn't looking too good right now either.

**Comment by warsaw** [comment]
This is simply an incorrect reading of history, although it can be understood in the context of the many different revision control systems python-mode.el has been maintained under.  Tim Peters was the original author of python-mode.el in 1992 (2 years before I even heard about Python).  If I remember correctly, he maintained it under RCS or CVS originally.  When I began to help Tim in 1994 or so, and then took over in 1995, I believe we eventually migrated the VCS to Subversion, then bazaar, then finally git.  It cannot be surprising that all those VCS migrations didn't preserve history to a very high fidelity.  Those are the historical facts.
